src/Indicator/PhpUnitTestCaseIndicator.php

```diff
@@ -51,24 +51,54 @@ public function isPhpUnitClass(Tokens $tokens, $index)
 
     /**
      * @param Tokens $tokens
+     * @param bool   $beginAtBottom whether we should start yielding phpunit classes from the bottom of the file
      *
      * @return \Generator array of [int start, int end] indexes from sooner to later classes
      */
-    public function findPhpUnitClasses(Tokens $tokens)
+    public function findPhpUnitClasses(Tokens $tokens, $beginAtBottom = true)
+    {
+        if (!$beginAtBottom) {
+            return $this->findPhpUnitClassesStartingFromTop($tokens);
+        }
+
+        return $this->findPhpUnitClassesStartingAtBottom($tokens);
+    }
+
+    /**
+     * @param Tokens $tokens
+     *
+     * @return \Generator
+     */
+    private function findPhpUnitClassesStartingFromTop(Tokens $tokens)
     {
         for ($index = 0, $limit = $tokens->count() - 1; $index < $limit; ++$index) {
-            if (!$tokens[$index]->isGivenKind(T_CLASS)) {
+            if (!$tokens[$index]->isGivenKind(T_CLASS) || !$this->isPhpUnitClass($tokens, $index)) {
                 continue;
             }
 
-            $classIndex = $index;
             $index = $tokens->getNextTokenOfKind($index, ['{']);
             $endIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $index);
-            if ($this->isPhpUnitClass($tokens, $classIndex)) {
-                yield [$index, $endIndex];
-            }
+            yield [$index, $endIndex];
 
             $index = $endIndex;
         }
     }
+
+    /**
+     * @param Tokens $tokens
+     *
+     * @return \Generator
+     */
+    private function findPhpUnitClassesStartingAtBottom(Tokens $tokens)
+    {
+        for ($index = $tokens->count() - 1, $limit = 0; $index > $limit; --$index) {
+            if (!$tokens[$index]->isGivenKind(T_CLASS) || !$this->isPhpUnitClass($tokens, $index)) {
+                continue;
+            }
+
+            $startIndex = $tokens->getNextTokenOfKind($index, ['{']);
+            $endIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $startIndex);
+            yield [$startIndex, $endIndex];
+        }
+    }
 }
```

Comment:
I think `findPhpUnitClassesStartingFromTop` and `findPhpUnitClassesStartingAtBottom` can unified into something like;

```php
    /**
     * @param Tokens $tokens
     * @param int    $direction 1|-1
     *
     * @return \Generator
     */
    private function findPhpUnitClasses(Tokens $tokens, $direction)
    {
        $index = 1 === $direction ? 0 : $tokens->count() - 1;
        while (true) {
            if (!$tokens[$index]->isGivenKind(T_CLASS) || !$this->isPhpUnitClass($tokens, $index)) {
                continue;
            }

            $startIndex = $tokens->getTokenOfKindSibling($index, $direction, ['{'], false);
            if (null === $startIndex) {
                return;
            }

            $endIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $startIndex);
            yield [$startIndex, $endIndex];

            $index += $direction;
        }
    }
```

Reply:
almost,
`$tokens->getTokenOfKindSibling($index, $direction, ['{'])` is wrong, as `{` will always be after `T_CLASS`, never before.

I will commit the update to the PR

Reply:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3745/commits/d4c609323be6bc82b084d5511be81f1e2991fef0

Reply:
```php
    public function findPhpUnitClasses(Tokens $tokens, $beginAtBottom = true)
    {
        $direction = $beginAtBottom ? -1 : 1;
        for ($index = 1 === $direction ? 0 : $tokens->count() - 1; $tokens->offsetExists($index); $index += $direction) {
            if (!$tokens[$index]->isGivenKind(T_CLASS) || !$this->isPhpUnitClass($tokens, $index)) {
                continue;
            }
            $startIndex = $tokens->getNextTokenOfKind($index, ['{'], false);
            if (null === $startIndex) {
                return;
            }
            $endIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $startIndex);
            yield [$startIndex, $endIndex];
        }
    }
```

Reply:
thanks @keradus :+1: